Tidy up formatting of type aliases in docs

pmenczel opened the pull request:
Following the workaround in [this stackoverflow answer](https://stackoverflow.com/a/67483317), added `autodoc_type_aliases` to the config of sphinx and added `from __future__ import annotations` where necessary. Makes it so that the docs say
```
H: QobjEvoLike
```
instead of
```
H: Qobj | QobjEvo | ~qutip.typing.QEvoProtocol | tuple['Qobj', typing.Union[ForwardRef('Coefficient'), str, qutip.typing.CoeffProtocol, numpy.ndarray, scipy.interpolate._interpolate.PPoly, scipy.interpolate._bsplines.BSpline, typing.Any]] | ~typing.Sequence[~qutip.typing.QEvoProtocol | Qobj | tuple['Qobj', typing.Union[ForwardRef('Coefficient'), str, qutip.typing.CoeffProtocol, numpy.ndarray, scipy.interpolate._interpolate.PPoly, scipy.interpolate._bsplines.BSpline, typing.Any]]]
```

(Also cleaned up imports in files I touched, and fixed a typo in Qobj type annotations.)

Ericgig commented:
Do we need the `from __future__ import annotations`? I used a lot of python 3.10 feature when adding types hints and removed support for python 3.9. 

Also should we add `ArrayLike`? We use it a few times such as for `tlist` in solvers:

sesolve(H: QobjEvoLike, psi0: [Qobj](https://github.com/qutip/qutip/pull/classes.html#qutip.core.qobj.Qobj), tlist: _SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes], e_ops: dict[Any, Union[[qutip.core.qobj.Qobj](https://github.com/qutip/qutip/pull/classes.html#qutip.core.qobj.Qobj), [qutip.core.cy.qobjevo.QobjEvo](https://github.com/qutip/qutip/pull/classes.html#qutip.core.cy.qobjevo.QobjEvo), Callable[[float, [qutip.core.qobj.Qobj](https://github.com/qutip/qutip/pull/classes.html#qutip.core.qobj.Qobj)], Any]]] = None, args: dict[str, Any] = None, options: dict[str, Any] = None, **kwargs)→ [Result](https://github.com/qutip/qutip/pull/classes.html#qutip.solver.result.Result)

pmenczel commented:
> Do we need the `from __future__ import annotations`? I used a lot of python 3.10 feature when adding types hints and removed support for python 3.9.

Yes. At least in my python 3.12 environment, the `autodoc_type_aliases` has no effect otherwise.

> Also should we add `ArrayLike`?

You are right, added it.
